pkg/koordlet/util/resctrl/resctrl.go

```diff
@@ -0,0 +1,321 @@
+	R.l.RLock()
+	defer R.l.RUnlock()
+
+	if v, ok := R.Apps[id]; ok {
+		return v, true
+	} else {
+		return App{}, false
+	}
+}
+
+func (R *RDTEngine) calculateMba(mbaPercent int64) int64 {
+	if R.Vendor == sysutil.INTEL_VENDOR_ID {
+		return calculateIntelMba(mbaPercent)
+	} else if R.Vendor == sysutil.AMD_VENDOR_ID {
+		return calculateAMDMba(mbaPercent)
+	}
+	return 0
+}
+
+func calculateIntelMba(mbaPercent int64) int64 {
+	if mbaPercent%10 != 0 {
+		actualPercent := mbaPercent/10*10 + 10
+		klog.V(4).Infof("cat MBA must multiple of 10, mbaPercentConfig is %d, actualMBAPercent will be %d",
+			mbaPercent, actualPercent)
+		return actualPercent
+	}
+
+	return mbaPercent
+}
+
+func calculateAMDMba(mbaPercent int64) int64 {
+	if mbaPercent == 100 {
+		return AMDCCDUnlimitedMB
+	}
+	mbaLimitValue := float64(AMDCCDMaxMBGbps*mbaPercent) / 100
+	return int64(mbaLimitValue)
+}
+
+func (R *RDTEngine) ParseSchemata(config apiext.ResctrlConfig, cbm uint) *sysutil.ResctrlSchemataRaw {
+	ids, _ := sysutil.CacheIdsCacheFunc()
+	schemataRaw := sysutil.NewResctrlSchemataRaw(ids).WithL3Num(len(ids))
+	if config.MB.Schemata.Percent != 0 {
+		percent := R.calculateMba(int64(config.MB.Schemata.Percent))
+		for k := range schemataRaw.MB {
+			schemataRaw.MB[k] = percent
+		}
+	}
+
+	if config.MB.SchemataPerCache != nil {
+		for _, v := range config.MB.SchemataPerCache {
+			percent := R.calculateMba(int64(v.Percent))
+			schemataRaw.MB[v.CacheID] = percent
+		}
+	}
+
+	if config.LLC.Schemata.Range != nil && len(config.LLC.Schemata.Range) == 2 {
+		start := config.LLC.Schemata.Range[0]
+		end := config.LLC.Schemata.Range[1]
+
+		l3MaskValue, err := sysutil.CalculateCatL3MaskValue(cbm, int64(start), int64(end))
+		if err != nil {
+			klog.Warningf("failed to calculate l3 cat schemata err: %v", err)
+			return schemataRaw
+		}
+
+		schemataRaw.WithL3Num(len(ids)).WithL3Mask(l3MaskValue)
+	}
+
+	if config.LLC.SchemataPerCache != nil {
+		for _, v := range config.LLC.SchemataPerCache {
+			if len(v.Range) == 2 {
+				start := v.Range[0]
+				end := v.Range[1]
+				l3MaskValue, err := sysutil.CalculateCatL3MaskValue(cbm, int64(start), int64(end))
+				if err != nil {
+					klog.Warningf("failed to calculate l3 cat schemata err: %v", err)
+					return schemataRaw
+				}
+				// l3 mask MUST be a valid hex
+				maskValue, err := strconv.ParseInt(strings.TrimSpace(l3MaskValue), 16, 64)
+				if err != nil {
+					klog.V(5).Infof("failed to parse l3 mask %s, err: %v", l3MaskValue, err)
+				}
+				schemataRaw.L3[v.CacheID] = maskValue
+			}
+		}
+	}
+	return schemataRaw
+}
+
+func GetPodCgroupNewTaskIdsFromPodCtx(podMeta *protocol.PodContext, tasksMap map[int32]struct{}) []int32 {
+	var taskIds []int32
+
+	for containerId, v := range podMeta.Request.ContainerTaskIds {
+		containerDir, err := koordletutil.GetContainerCgroupParentDirByID(podMeta.Request.CgroupParent, containerId)
+		if err != nil {
+			klog.Errorf("container %s lost during reconcile", containerDir)
```

Comment:
To distinguish the cgroup deleted error and skip, it is recommended to check if the error type is NotFound and log with V(5).